Contribution Guidelines Update 🧽 (#666)
primer/components: CONTRIBUTING.md

```
yarn run lint
```

Or, you can use [npx] to run eslint on one or more specific files:

```sh
# lint the component and the tests in src/__tests__
npx eslint src/**/MyComponent.js
```

**Protip:** The [eslint `--fix` flag](https://eslint.org/docs/user-guide/command-line-interface#--fix) can automatically fix most linting errors, such as those involving whitespace or incorrect ordering of object keys and imports. You can fix those issues across the entire project with:

```sh
yarn run lint -- --fix
```

**Protip:** `yarn run lint -- --quiet` (or `npx eslint --quiet ...`) will suppress warnings so that you can focus on fixing errors.

### Testing

We test our components with [Jest](https://facebook.github.io/jest/) and [react-test-renderer](https://reactjs.org/docs/test-renderer.html). You can run the tests locally with `yarn test`. To run the tests as you work, run Jest in watch mode with:

```
yarn test -- --watch
```

See [`src/__tests__/example.js`](src/__tests__/example.js) for examples of ways that we test our components.

### TypeScript support

> Several of the projects that consume Primer Components are written in TypeScript! Though Primer Components is not currently written in TS, we do export type definitions in order to make Primer Components compatible with other TS projects.

Reviewer: The `!` seems like overkill here.
```suggestion
Several of the projects that consume Primer Components are written in TypeScript. Though Primer Components is not currently written in TS, we do export type definitions in order to make Primer Components compatible with other TS projects.
```

Whenever adding new components or modifying the props of an existing component, **please make sure to update the type definition** in `index.d.ts`! This is super important to make sure we don't break compatibility :) 

### Additional resources

- [Primer Components Philosophy](https://primer.style/components/philosophy)
- [Primer Components Core Concepts](https://primer.style/components/core-concepts)
- [Primer Components System Props](https://primer.style/components/system-props)
- [Styled Components docs](https://styled-components.com/)
- [Styled System docs](https://styled-system.com/)

## Writing documentation

We use [Doctocat](https://github.com/primer/doctocat) to power our documentation site at [https://primer.style/components](https://primer.style/components/).

> To add a new component to our documentation site, create a new file with the `.md` extension for your component in `docs/content/components`.

Reviewer: The path is actually `docs/content`. I think we did this to avoid having paths like `primer.style/components/components/Button`

```suggestion
To add a new component to our documentation site, create a new file with the `.md` extension for your component in `docs/content` (e.g. `docs/content/Button.md`).
```

## Creating a pull request

> When you're ready to create a pull request, open one using the pull request template. Be sure to add screenshots of any relevant work and a thoughtful description.

Reviewer: It's a little unclear _how_ to use the pull request template from this sentence. Is the pull request template automatically populated when you open a PR?

### What to expect after opening a pull request
After opening a pull request, a member of the design systems team will add the appropriate labels (major, minor, patch release labels) and update the base branch to the correct release branch. Usually, you'll receive a response from the design systems team within a day or two. The design systems team member will review the pull request keeping the following items in mind:

### What we look for in reviews

> - If it's a new component, does the component make sense to add to Primer Components? (Ideally this is discussed before the pull request stage, please reach out to a DS member if you aren't sure if a component should be added to Primer Componnets!)

Reviewer: ```suggestion
- If it's a new component, does the component make sense to add to Primer Components? (Ideally this is discussed before the pull request stage, please reach out to a DS member if you aren't sure if a component should be added to Primer Components)
```

> - Does the component follow our Primer Components code style?

Reviewer: Do we have the code style documented? If so, we might want to link to it here.

- Does the component use theme values for most CSS values?
- Does the component have access to the [default theme](#adding-default-theme)?
- Does the component have the [correct system props implemented](#adding-system-props)?
- Is the component API intuitive?
- Does the component have the appropriate [type definitions in `index.d.ts`](#typescript-support)?
- Is the component documented accurately?
- Does the component have appropriate tests?

> - Does the pull request increase the bundle size significantly?

Reviewer: I'd love to have some automated way to enforce a small bundle size. @siddharthkp's bundlesize project looks promising: https://github.com/siddharthkp/bundlesize

Author: Yes! I was thinking that too

Reviewer 2: I was thinking of that too 😄

## Deploying and publishing

### Deploying
All of our documentation sites use the [Now integration](https://github.com/organizations/primer/settings/installations/1007619) to deploy documentation changes whenever code is merged into master. The integration also creates a preview site every time you commit code to a branch. To view the preview site, navigate to the PR and find the comment from the `now` bot. This will include a link to the preview site for your branch. 

Once you merge your branch into master, any changes to the docs will automatically deploy. No further action is necessary.

### Path aliasing
This site is served as a subdirectory of [primer.style] using a [path alias](https://zeit.co/docs/features/path-aliases) configured in that repo's [`rules.json`](https://github.com/primer/primer.style/tree/master/rules.json). If you change the production deployment URL for this app, you will also need to change it there and re-deploy that app; otherwise, Now will automatically route requests from [primer.style/components](https://primer.style/components/) to the new deployment whenever you alias this one to `primer-components.now.sh`.

### Publishing

> We use a custom GitHub Actiont to handle all of our processes relating to publishing to NPM. This includes release candidates, canary releases, and publishing the final release.

Reviewer: ```suggestion
We use a custom GitHub Actions to handle all of our processes relating to publishing to NPM. This includes release candidates, canary releases, and publishing the final release.
```
